src/main/java/com/conveyal/datatools/manager/controllers/api/ServerController.java

```diff
@@ -0,0 +1,343 @@
+                validateSubnetId(server.ec2Info.subnetId, req);
+                validateSecurityGroupId(server.ec2Info.securityGroupId, req);
+                validateIamRoleArn(server.ec2Info.iamRoleArn, req);
+                validateKeyName(server.ec2Info.keyName, req);
+                validateAmiId(server.ec2Info.amiId, req);
+                if (server.ec2Info.instanceCount < 0) server.ec2Info.instanceCount = 0;
+            }
+            // Server must have name.
+            if (isEmpty(server.name))
+                logMessageAndHalt(req, HttpStatus.BAD_REQUEST_400, "Server must have valid name.");
+            // Server must have an internal URL (for build graph over wire) or an s3 bucket (for auto deploy ec2).
+            if (isEmpty(server.s3Bucket)) {
+                if (server.internalUrl == null || server.internalUrl.size() == 0) {
+                    logMessageAndHalt(req, HttpStatus.BAD_REQUEST_400, "Server must contain either internal URL(s) or s3 bucket name.");
+                }
+            } else {
+                // Verify that application has permission to write to/delete from S3 bucket. We're following the recommended
+                // approach from https://stackoverflow.com/a/17284647/915811, but perhaps there is a way to do this
+                // effectively without incurring AWS costs (although writing/deleting an empty file to S3 is probably
+                // miniscule).
+                String key = UUID.randomUUID().toString();
+                try {
+                    FeedStore.s3Client.putObject(server.s3Bucket, key, File.createTempFile("test", ".zip"));
+                    FeedStore.s3Client.deleteObject(server.s3Bucket, key);
+                } catch (IOException | AmazonS3Exception e) {
+                    String message = "Cannot write to specified S3 bucket " + server.s3Bucket;
+                    LOG.error(message, e);
+                    logMessageAndHalt(req, 400, message, e);
+                }
+            }
+        } catch (Exception e) {
+            if (e instanceof HaltException) throw e;
+            else logMessageAndHalt(req, 400, "Error encountered while validating server field", e);
+        }
+    }
+
+    private static void validateAmiId(String amiId, Request req) {
+        String message = "Server must have valid AMI ID (or field must be empty)";
+        if (isEmpty(amiId)) return;
+        if (!amiExists(amiId)) logMessageAndHalt(req, HttpStatus.BAD_REQUEST_400, message);
+    }
+
+    public static boolean amiExists(String amiId) {
+        try {
+            DescribeImagesRequest request = new DescribeImagesRequest().withImageIds(amiId);
+            DescribeImagesResult result = ec2.describeImages(request);
+            // Iterate over AMIs to find a matching ID.
+            for (Image image : result.getImages()) if (image.getImageId().equals(amiId)) return true;
+        } catch (AmazonEC2Exception e) {
+            LOG.info("AMI does not exist.", e);
```

Comment:
Could it be possible that an error prevented the proper checking of the AMI? I think the error message should be changed to reflect that possibility.